efficientdet/keras/train_lib.py

```diff
@@ -0,0 +1,477 @@
+      per_replica_loss = self.distribute_strategy.run(func, args=batch)
+      total_loss += self.distribute_strategy.reduce(tf.distribute.ReduceOp.MEAN,
+                                                    per_replica_loss,
+                                                    axis=None)
+      [callback.on_batch_end(step, logs) for callback in callbacks]
+      total_loss /= step
+      tf.print('\r', step, ':', total_loss, end='', sep='')
+
+  def _dataset_loop(self, dataset_iter, func, logs, callbacks):
+    total_loss = 0.0
+    step = 1
+    for batch in dataset_iter:
+      self.step.assign(step)
+      [callback.on_batch_begin(step, logs) for callback in callbacks]
+      per_replica_loss = self.distribute_strategy.run(func, args=batch)
+      total_loss += self.distribute_strategy.reduce(tf.distribute.ReduceOp.SUM,
+                                                    per_replica_loss,
+                                                    axis=None)
+      [callback.on_batch_end(step, logs) for callback in callbacks]
+      total_loss /= step
+      tf.print('\r', step, ':', total_loss, end='', sep='')
+      step += 1
+
+  @tf.function
+  def train_step(self, features, labels):
+    with tf.GradientTape() as tape:
+      cls_outputs, box_outputs = self(features, training=True)
+      det_loss, cls_loss, box_loss, box_iou_loss = self._detection_loss(
+          cls_outputs, box_outputs, labels)
+      reg_l2loss = self._reg_l2_loss(self.config.weight_decay)
+      total_loss = det_loss + reg_l2loss
+      if isinstance(self.optimizer,
+                    tf.keras.mixed_precision.experimental.LossScaleOptimizer):
+        scaled_loss = self.optimizer.get_scaled_loss(total_loss)
+      else:
+        scaled_loss = total_loss
+    trainable_vars = self.trainable_variables
+    scaled_gradients = tape.gradient(scaled_loss, trainable_vars)
+    if isinstance(self.optimizer,
+                  tf.keras.mixed_precision.experimental.LossScaleOptimizer):
+      gradients = self.optimizer.get_unscaled_gradients(scaled_gradients)
+    else:
+      gradients = scaled_gradients
+    if self.config.clip_gradients_norm > 0:
+      gradients, gnorm = tf.clip_by_global_norm(gradients,
+                                                self.config.clip_gradients_norm)
+    optimizer_op = self.optimizer.apply_gradients(zip(gradients,
+                                                      trainable_vars))
+    if self.config.moving_average_decay:
+      with tf.control_dependencies([optimizer_op]):
+        self.ema.apply(self.trainable_variables)
+    return total_loss
+
+  @tf.function
+  def test_step(self, features, labels):
+    cls_outputs, box_outputs = self(features, training=False)
+    det_loss, cls_loss, box_loss, box_iou_loss = self._detection_loss(
+        cls_outputs, box_outputs, labels)
+    reg_l2loss = self._reg_l2_loss(self.config.weight_decay)
+    total_loss = det_loss + reg_l2loss
+    return total_loss
+
+  def train(self,
```

Comment:
Should we use model.fit/evaluate according to https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit?

Reply:
As discussed in the other comment, let's just use train in this PR.

Could you help add some doc or link for this train function? (looks like it is not a member of tf.keras.Model?)  
Also, it would be great if we can somehow add a simple test to train_lib_test.py